### PR TITLE
Use less stack in mdns, have more const correctness

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
Mdns platform code modifies protocol buffers in-place and this resulted in hard to spot bugs. Also uses strncpy which is error prone (uses it correctly, but code is longer as a result)

 #### Summary of Changes
  - separate out the 'extract protocol' from 'truncate type' methods , to make the code easier to understand
  - all methods now are const-correct and cannot buffer overflow
  - use the chip platform copy string methods as a replacement for strncpy

### CODE REVIEW

File: src/platform/Linux/MdnsImpl.cpp [hunk 1]
```diff
@@ -540,25 +541,40 @@ CHIP_ERROR MdnsAvahi::Browse(const char * type, MdnsServiceProtocol protocol, ch
...
+    {
+        return MdnsServiceProtocol::kMdnsProtocolUdp;
+    }
+
+    ChipLogError(Discovery, "Unknown protocol in type: %s", type);
+    return MdnsServiceProtocol::kMdnsProtocolUnknown;
+}
+
+template <size_t N>
+void CopyWithoutProtocol(char (&dest)[N], const char * typeAndProtocol)
```
Comment (wgtdkp) [inline src/platform/Linux/MdnsImpl.cpp:568]: nit:
```suggestion
void CopyTypeWithoutProtocol(char (&dest)[N], const char * typeAndProtocol)
```

**Review by wgtdkp — APPROVED** [review]
Looks Great! Thanks @andy31415

File: src/platform/Linux/MdnsImpl.cpp [hunk 2]
```diff
@@ -540,25 +541,40 @@ CHIP_ERROR MdnsAvahi::Browse(const char * type, MdnsServiceProtocol protocol, ch
...
+{
+    const char * dotPos          = strrchr(typeAndProtocol, '.');
+    size_t lengthWithoutProtocol = (dotPos != nullptr) ? static_cast<size_t>(dotPos - typeAndProtocol) : N;
+
+    Platform::CopyString(dest, typeAndProtocol);
+    if (lengthWithoutProtocol < N)
+    {
+        dest[lengthWithoutProtocol] = 0;
```
Comment (tcarmelveilleux) [inline src/platform/Linux/MdnsImpl.cpp:576]: Why must you still maintain the null-termination invariant yourself?

The CopyString will actually do the null-termination ehre even if lengthWithoutProtocol == N, due to the call `Platform::CopyString(dest, typeAndProtocol)` --> `Platform::CopyString(dest, typeAndProtocol,N)` internal expansion:

```
inline char * CopyString(char * dest, const char * source, size_t length)
{
    if (dest)
    {
        strncpy(dest, source, length);
        dest[length] = 0;
    }
    return dest;
}

/**
 * Convenience method for CopyString to auto-detect destination size.
 */
template <size_t N>
inline char * CopyString(char (&dest)[N], const char * source)
{
    return CopyString(dest, source, N);
}
```

Comment (andy31415) [reply src/platform/Linux/MdnsImpl.cpp:576]: I need to truncate stuff after the .

Comment (andy31415) [reply src/platform/Linux/MdnsImpl.cpp]: Updated with some renames and comments.

The CopyString method was buggy: it set a array[sizeof(array)] to 0 which is bad. Fixed that too - thank you for catching.

File: src/lib/support/CHIPMemString.h [hunk 3]
```diff
@@ -82,7 +82,7 @@ inline char * CopyString(char (&dest)[N], const char * source)
  */
 inline char * MemoryAllocString(const char * string, size_t length)
 {
-    return CopyString(static_cast<char *>(MemoryAlloc(length + 1)), string, length);
+    return CopyString(static_cast<char *>(MemoryAlloc(length + 1)), length + 1, string);
```
Comment (tcarmelveilleux) [inline src/lib/support/CHIPMemString.h]:
```suggestion
    size_t length_with_null = length + 1;
    return CopyString(static_cast<char *>(MemoryAlloc(length_with_null)), length_with_null, string);
```

File: src/platform/Linux/MdnsImpl.cpp (hunk 1 above)
Comment (andy31415) [reply src/platform/Linux/MdnsImpl.cpp]: Updated.

File: src/platform/Linux/MdnsImpl.cpp [hunk 4]
```diff
@@ -540,25 +541,45 @@ CHIP_ERROR MdnsAvahi::Browse(const char * type, MdnsServiceProtocol protocol, ch
...
+    return MdnsServiceProtocol::kMdnsProtocolUnknown;
+}
+
+/// Copies the type from a string containing both type and protocol
+///
+/// e.g. if input is "foo.bar", output is "foo", input is 'a.b._tcp", oputput is "a.b"
+template <size_t N>
+void TypeFromTypeAndProtocol(char (&dest)[N], const char * typeAndProtocol)
+{
+    const char * dotPos          = strrchr(typeAndProtocol, '.');
```
Comment (tcarmelveilleux) [inline src/platform/Linux/MdnsImpl.cpp:573]: static_assert(N > 0)

**Review by tcarmelveilleux — APPROVED** [review]

File: src/platform/Linux/MdnsImpl.cpp (hunk 2 above)
Comment (andy31415) [reply src/platform/Linux/MdnsImpl.cpp]: Did a LOT of yak shaving here to make the CopyString seem a bit safer. I dislike the idea of 'pass in source length' and rather find 'pass in destination available size' to be a least unexpected behaviour.

Hoping code health is improved now. Also a bit frustrated that ChipMemStr has no unit tests.

File: src/lib/support/CHIPMemString.h (hunk 3 above)
Comment (andy31415) [reply src/lib/support/CHIPMemString.h]: done.

File: src/platform/Linux/MdnsImpl.cpp (hunk 4 above)
Comment (andy31415) [reply src/platform/Linux/MdnsImpl.cpp:573]: Is this needed? we are very unlikely to create and pass in arrays of size 0.

Comment (andy31415) [reply src/platform/Linux/MdnsImpl.cpp:573]: Also assuming size 0 ... I think code will still be correct (but nothing useful will come out of it as even null terminator will be missing)

File: src/platform/Linux/MdnsImpl.cpp [hunk 5]
```diff
@@ -540,25 +541,45 @@ CHIP_ERROR MdnsAvahi::Browse(const char * type, MdnsServiceProtocol protocol, ch
...
+        return MdnsServiceProtocol::kMdnsProtocolUdp;
+    }
+
+    ChipLogError(Discovery, "Unknown protocol in type: %s", type);
+    return MdnsServiceProtocol::kMdnsProtocolUnknown;
+}
+
+/// Copies the type from a string containing both type and protocol
+///
+/// e.g. if input is "foo.bar", output is "foo", input is 'a.b._tcp", oputput is "a.b"
```
Comment (kpschoedel) [inline src/platform/Linux/MdnsImpl.cpp:569]: o~p~utput (Although ‘oputput’ is not bad; I could also accept replacing ‘input’ with ‘igetget’.)

**Review by kpschoedel — APPROVED** [review]

**Comment by andy31415** [comment]
@woody-apple @bzbarsky-apple @saurabhst @msandstedt @jmartinez-silabs ?
Trying to get this in because fixes in CopyString

**Review by saurabhst — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
@msandstedt @mspang ?

**Review by jmartinez-silabs — APPROVED** [review]

**Pull request merged**